Meter.c

```diff
@@ -276,43 +298,869 @@ static void BarMeterMode_draw(Meter* this, int x, int y, int w) {
+      return 0;
+
+   assert(stack->valueSum <= DBL_MAX);
+   double stackArea = (stack->valueSum / scaledTotal) * (double)(int)graphHeight;
+   double adjNCells = adjOffset ? (double)(int)adjOffset->nCells : 0.0;
+
+   if (stackArea - adjNCells > (halfPoint - stack->startPoint) * 2.0) {
+      // Mathematically equivalent to:
+      // (stack->startPoint + stackArea / 2 > halfPoint + adjNCells / 2)
+      return 1;
+   }
+
+   if (stackArea - adjNCells < (halfPoint - stack->startPoint) * 2.0)
+      return 0;
+
+   return -1;
+}
+
+static void GraphMeterMode_addItemAdjOffset(GraphColorAdjOffset* adjOffset, unsigned int nCells) {
+   adjOffset->offsetVal += (uint32_t)adjOffset->nCells * 2 + nCells;
+   adjOffset->nCells += nCells;
+}
+
+static void GraphMeterMode_addItemAdjStack(GraphColorAdjStack* stack, double scaledTotal, double value) {
+   assert(scaledTotal <= DBL_MAX);
+   assert(stack->valueSum < DBL_MAX);
+
+   stack->fractionSum += (stack->valueSum / scaledTotal) * 2.0;
+   stack->valueSum += value;
+
+   assert(stack->nItems < UINT8_MAX);
+   stack->nItems++;
+}
+
+static uint16_t GraphMeterMode_makeDetailsMask(const GraphColorComputeState* prev, const GraphColorComputeState* new, double rem, int blanksAtTopCell) {
+   assert(new->nCellsPainted > prev->nCellsPainted);
+   assert(rem >= 0.0);
+   assert(rem < 1.0);
+
+   bool isItemOnEdge = (blanksAtTopCell >= 0 || prev->nCellsPainted == 0);
+   const uint8_t dotAlignment = 2;
+   if (isItemOnEdge && rem > 0.0 && rem < (0.125 * dotAlignment))
+      rem = (0.125 * dotAlignment);
+
+   uint8_t blanksAtEnd;
+   bool roundsUpInAscii = false;
+   bool roundsDownInAscii = false;
+   if (blanksAtTopCell >= 0) {
+      assert(blanksAtTopCell < 8);
+      blanksAtEnd = (uint8_t)blanksAtTopCell;
+      roundsUpInAscii = true;
+   } else if (prev->nCellsPainted == 0 || prev->topPoint <= (double)(int)prev->nCellsPainted) {
+      blanksAtEnd = (uint8_t)((1.0 - rem) * 8.0) % 8;
+   } else if ((double)(int)new->nCellsPainted > new->topPoint) {
+      assert((int)new->nCellsPainted - new->topPoint < 1.0);
+      // Unlike other conditions, this one rounds to nearest for visual reason.
+      // In case of a tie, display the dot at lower position of the graph,
+      // i.e. MSB of the "details" data.
+      double fraction = (1.0 - rem) * 8.0;
+      fraction -= (int)fraction;
+      double distance = ((double)(int)new->nCellsPainted - new->topPoint) * 8.0 + (1.0 - fraction) * 0.5;
+      blanksAtEnd = (uint8_t)distance;
+
+      // Tiebreaking direction that may be needed in the ASCII display mode.
+      distance = distance - (int)distance;
+      assert(distance >= 0.0);
+      assert(distance < 1.0);
+      roundsUpInAscii = distance < 0.5;
+      roundsDownInAscii = distance > 0.5;
+   } else {
+      blanksAtEnd = 0;
+   }
+   assert(blanksAtEnd < 8);
+
+   uint8_t blanksAtStart;
+   if (prev->nCellsPainted > 0) {
+      blanksAtStart = (uint8_t)((1.0 - rem) * 8.0) % 8 - blanksAtEnd;
```

Comment:
I hit the assert() below with the MemoryMeter.
Maybe use
```c
blanksAtStart = (uint8_t)((1.0 - rem) * 8.0) % 8;
if (blanksAtStart > blanksAtEnd)
   blanksAtStart -= blanksAtEnd;
else
   blanksAtStart = 0;
```
?

Reply:
Which assertion error did you hit in particular?
Besides, I am partly rewriting the algorithm so that your suggestion might not be applicable.

`blanksAtStart` shouldn't wrap to a negative value and ideally there's no need to clamp it.

Reply:
The `assert(blanksAtStart < 8);` couple lines below

Reply:
Oh wait. I was thinking a rare case where `blanksAtStart` could wrap to below zero...